### PR TITLE
[5.0] Add touchOwners method to DB/Eloquent/Collection

### LINKED ISSUE #6386: $touches causes fatal error if no ManyToMany relations exist
You can set up a pair of Eloquent models like this:

    class Product extends \Eloquent {
        public function assets() {
            return $this->belongsToMany('Asset');
        }
    }

    class Asset extends \Eloquent {
        protected $touches = [ 'products' ];

        public function products() {
            return $this->belongsToMany('Product');
        }
    }

By doing this, you can call

    Product::find( $validProductId )->assets()->attach( $validAssetId );

the `Asset` will go and update the `updated_at` field on every related `Product` to indicate that its relations have changed.

However, if you make a new `Asset`, or take an `Asset` which has no related `Product`s yet, you'll get a fatal error as the `Model::touchOwners` method only checks if `$this->relation` isn't null:

	public function touchOwners()
	{
		foreach ($this->touches as $relation)
		{
			$this->$relation()->touch();

			if ( ! is_null($this->$relation)) // this is the problem
			{
				$this->$relation->touchOwners();
			}
		}
	}

I have a tentative fix in place locally (for now) by just adding in a check for `&& ! $this->$relation->isEmpty()`, but that may affect other relations, I'm not sure:

	public function touchOwners()
	{
		foreach ($this->touches as $relation)
		{
			$this->$relation()->touch();

			if ( ! is_null($this->$relation) && ! $this->$relation->isEmpty())
			{
				$this->$relation->touchOwners();
			}
		}
	}

### PR DESCRIPTION
**Pull request by impleri**
Fixes #6386 in 5.0.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Let's just check if the relation is empty as was suggested instead of adding this to the collection.

**Comment by taylorotwell** [comment]
Also if transform is an array_map shouldn't you be returning a value in your Closure?

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 1]
```diff
@@ -132,6 +132,16 @@ public function modelKeys()
 	}
 
 	/**
+	 * Touch the models in the collection.
+	 *
+	 * @return void
+	 */
+	public function touchOwners()
+	{
+		$this->transform(function($m) { $m->touch(); });
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: Please try to avoid singe line statements like this. They're fine for the tests, but not for src.

Comment (impleri) [inline src/Illuminate/Database/Eloquent/Collection.php]: Sure can. Was trying to match the styling as in modelKeys()

```php
public function modelKeys()
{
    return array_map(function($m) { return $m->getKey(); }, $this->items);
}
```

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: @impleri Could you format it like this please: #8173.

**Comment by taylorotwell** [comment]
I'm just a little nervous this hasn't even been tested since the transform had absolutely no return.

**Comment by impleri** [comment]
@taylorotwell I'm testing a few scenarios first as per your first suggestion (check if empty and perhaps not add this to Collection). I'll update the PR ASAP. I'm just lagging behind.

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: @impleri Don't bother changing this btw. Taylor is fine with oneline.

**Comment by taylorotwell** [comment]
Can be re-opened when ready to merge.

**Pull request closed without merging**